src/engines/julia/update_rules/cvi.jl

```diff
@@ -0,0 +1,356 @@
+end
+
+function ruleSPCVIIn1Factor(node_id::Symbol,
+                            msg_out::Message{<:FactorFunction, <:VariateType},
+                            msg_in::Message{<:Gaussian, Multivariate})
+
+    thenode = currentGraph().nodes[node_id]
+
+    η = deepcopy(naturalParams(msg_in.dist))
+    if thenode.online_inference == false
+        λ_init = deepcopy(η)
+    else
+        λ_init = deepcopy(naturalParams(thenode.q[1]))
+    end
+
+    logp_nc(z) = (thenode.dataset_size/thenode.batch_size)*logPdf(msg_out.dist, thenode.g(z))
+    λ = renderCVI(logp_nc,thenode.num_iterations,thenode.opt,λ_init,msg_in)
+
+    λ_message = λ.-η
+
+    if thenode.proper_message λ_message = convertToProperMessage(msg_in, λ_message) end
+    thenode.q = [standardDist(msg_in.dist,λ)]
+    if thenode.online_inference == false thenode.q_memory = deepcopy(thenode.q) end
+    return standardMessage(msg_in.dist,λ_message)
+end
+
+function ruleSPCVIOutNFactorNode(node_id::Symbol,
+                                 msg_out::Nothing,
+                                 msg_in::Message)
+
+    thenode = currentGraph().nodes[node_id]
+
+    sampl = thenode.g(sample(msg_in.dist))
+    if length(sampl) == 1
+        variate = Univariate
+    else
+        variate = Multivariate
+    end
+    return Message(variate,SetSampleList,node_id=node_id)
+
+end
+
+function ruleSPCVIInFactorX(node_id::Symbol,
+                            inx::Int64,
+                            msg_out::Message{<:FactorFunction, <:VariateType},
+                            msgs_in::Vararg{Message})
+
+    thenode = currentGraph().nodes[node_id]
+
+    local_messages, global_messages = [], []
+    local_inx, global_inx = [], []
+    for i=1:length(thenode.online_inference)
+        if thenode.online_inference[i]
+            push!(global_messages, msgs_in[i])
+            push!(global_inx, i)
+        else
+            push!(local_messages, msgs_in[i])
+            push!(local_inx, i)
+        end
+    end
+    msgs_list = [local_messages;global_messages]
+    inx_list = [local_inx;global_inx]
+
+    arg_sample = (z,j) -> begin
+        samples_in = []
+        for k=1:length(msgs_in)
+            if k==j
+                push!(samples_in,collect(Iterators.repeat([ z ], thenode.num_samples)))
+            else
+                push!(samples_in,sample(thenode.q_memory[k], thenode.num_samples))
+            end
+        end
+
+        return samples_in
+    end
+
+    λ_list = []
+
+    if thenode.infer_memory == 0
+        for j=1:length(msgs_list)
+            msg_in = msgs_list[j]
+            logp_nc(z) = sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
+            if thenode.online_inference[inx_list[j]]
+                λ_init = deepcopy(naturalParams(thenode.q[inx_list[j]]))
+                logp_nc(z) = (thenode.dataset_size/thenode.batch_size)*sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
```

Comment:
You overwrite the method here. This is undesirable. Julia throws a warning, i.e.
```julia
WARNING: Method definition (::ForneyLab.var"#logp_nc#364"{msg_out, arg_sample, inx_list, thenode, j})(Any) in module ForneyLab at /Users/apodusenko/.julia/dev/ForneyLab/src/engines/julia/update_rules/cvi.jl:129 overwritten at /Users/apodusenko/.julia/dev/ForneyLab/src/engines/julia/update_rules/cvi.jl:132.
  ** incremental compilation may be fatally broken for this module **
```

One way to circumvent this warning was adviced by @bvdmitri:
```julia
logp_nc = (z) -> sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
if thenode.online_inference[inx_list[j]]
    λ_init = deepcopy(naturalParams(thenode.q[inx_list[j]]))
    logp_nc = (z) -> (thenode.dataset_size/thenode.batch_size)*sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
else
    λ_init = deepcopy(naturalParams(msg_in.dist))
end
```
If we enter the `if statement,` two functions will be created. The first one will be handled by a garbage collector.

Reply:
Another option would be:

```julia
(logp_nc, λ_init) = if thenode.online_inference[inx_list[j]]
    λ_init = deepcopy(naturalParams(thenode.q[inx_list[j]]))
    logp_nc = (z) -> (thenode.dataset_size/thenode.batch_size)*sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
    (logp_nc, λ_init)
else
    λ_init = deepcopy(naturalParams(msg_in.dist))
    logp_nc = (z) -> sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
if thenode.online_inference[inx_list[j]]
    (logp_nc, λ_init)
end
```

`if` statement in Julia returns an actual last value from executed branch, so you can use it as ternary operator:

```julia
value = if some_condition
    0.0
else
    1.0
end
```

Reply:
I couldn't make it work without overwriting the method at first place. But I'll consider @bvdmitri's suggestions.